packages/shared-ux/file/types/index.d.ts

```diff
@@ -251,13 +251,22 @@ export interface FileKind {
    * Unique file kind ID
    */
   id: string;
+
   /**
-   * Maximum size, in bytes, a file of this kind can be.
+   * Maximum size, in bytes, used to enforce HTTP payload size limits as well as
+   * used for the `maxUploadSize` value, if `maxUploadSize` is not specified.
    *
    * @default 4MiB
    */
   maxSizeBytes?: number;
 
+  /**
+   * Max file contents size, in bytes. Can be customized per file using the
+   * {@link FileJSON} object. If not specified, the `maxSizeBytes` value will
+   * still be used to enforce the max file size limit.
+   */
+  maxUploadSize?: (file: FileJSON) => number;
```

Comment:
> I left the maxSizeBytes as-is for now, because it is actually used in two places: (1) in the Files client; (2) in HTTP routes. So, I didn't find an easy way to reuse it.

It would be really nice to have just one way to set this:

```ts
maxSizeBytes: number | (file: FileJSON) => number;
```

As far as I can tell the main blocker is using it in the HTTP router configuration. If we remove it from there we could add another conditional to the `try..catch` in the upload handler to map the `MaxByteSizeExceededError` to a `413` response. And then keep the rest of your code mostly as-is.

If you feel differently, I'll approve to unblock progress, but I don't think it is a massive change to what you've already got.

Reply:
Also, for `(file: FileJSON) => number;`, wherever it lives, we should consider adding a comment that it is only supported/used server side.

Reply:
I'm curious if we don't use it for `maxBytes` on the server routes, would should we set the `maxBytes` to. Or do we even need to set it, normally, I would expect the `maxBytes` would apply to normal HTTP requests, but ours is a streaming one. Maybe it computes the total size of the stream as well. WDYT, I guess if we enforce it and it has to be set, we can set it to Infinity, or something very large?

Reply:
That is a good question. I think it may have some default, but setting to Infinity with a comment that we enforce this elsewhere could be a good approach

Reply:
Done.